src/dos/cdrom_image.cpp

```diff
@@ -305,110 +573,205 @@ int CDROM_Interface_Image::GetTrack(int sector)
+	        track,
+	        raw ? "true":"false",
+	        sector,
+	        length);
+#endif
 	return tracks[track].file->read(buffer, seek, length);
 }
 
 void CDROM_Interface_Image::CDAudioCallBack(Bitu len)
 {
-	len *= 4;       // 16 bit, stereo
-	if (!len) return;
-	if (!player.isPlaying || player.isPaused) {
-		player.channel->AddSilence();
+	// Our member object "playbackRemaining" holds the
+	// exact number of stream-bytes we need to play before meeting the
+	// DOS program's desired playback duration in sectors. We simply
+	// decrement this counter each callback until we're done.
+	if (len == 0 || !player.isPlaying || player.isPaused) {
 		return;
 	}
-	
-	SDL_mutexP(player.mutex);
-	while (player.bufLen < (Bits)len) {
-		bool success;
-		if (player.targetFrame > player.currFrame)
-			success = player.cd->ReadSector(&player.buffer[player.bufLen], true, player.currFrame);
-		else success = false;
-		
-		if (success) {
-			player.currFrame++;
-			player.bufLen += RAW_SECTOR_SIZE;
-		} else {
-			memset(&player.buffer[player.bufLen], 0, len - player.bufLen);
-			player.bufLen = len;
-			player.isPlaying = false;
+
+	// determine bytes per request (16-bit samples)
+	const Bit8u channels = player.trackFile->getChannels();
+	const Bit8u bytes_per_request = channels * 2;
+	int total_requested = len * bytes_per_request;
+
+	while (total_requested > 0) {
+		int requested = total_requested;
+
+		// Every now and then the callback wants a big number of bytes,
+		// which can exceed our circular buffer. In these cases we need
+		// read through as many iteration of our circular buffer as needed.
+		if (total_requested > AUDIO_DECODE_BUFFER_SIZE) {
+			requested = AUDIO_DECODE_BUFFER_SIZE;
+			total_requested -= AUDIO_DECODE_BUFFER_SIZE;
 		}
-	}
-	SDL_mutexV(player.mutex);
-	if (player.ctrlUsed) {
-		Bit16s sample0,sample1;
-		Bit16s * samples=(Bit16s *)&player.buffer;
-		for (Bitu pos=0;pos<len/4;pos++) {
-#if defined(WORDS_BIGENDIAN)
-			sample0=(Bit16s)host_readw((HostPt)&samples[pos*2+player.ctrlData.out[0]]);
-			sample1=(Bit16s)host_readw((HostPt)&samples[pos*2+player.ctrlData.out[1]]);
-#else
-			sample0=samples[pos*2+player.ctrlData.out[0]];
-			sample1=samples[pos*2+player.ctrlData.out[1]];
-#endif
-			samples[pos*2+0]=(Bit16s)(sample0*player.ctrlData.vol[0]/255.0);
-			samples[pos*2+1]=(Bit16s)(sample1*player.ctrlData.vol[1]/255.0);
+		else {
+			total_requested = 0;
 		}
+
+		// Three scenarios in order of probabilty:
+		//
+		// 1. Consume: If our decoded circular buffer is sufficiently filled to
+		//	     satify the requested size, then feed the callback with
+		//	     the requested number of bytes.
+		//
+		// 2. Wrap: If we've decoded and consumed to edge of our buffer, then
+		//	     we need to wrap any remaining decoded-but-not-consumed
+		//	     samples back around to the front of the buffer.
+		//
+		// 3. Fill: When out circular buffer is too depleted to satisfy the
```

Comment:
Typo? `out` was probably `our`?

Reply:
559de7d454c79ea6d88420d000a51e170d1ac4daFixed in